applications/server/modules/access.py

```diff
@@ -0,0 +1,83 @@
+from gluon import current
+from gluon import auth
+
+
+# Permission types.
+# v = view
+# a = admin (valid only for one whole device)
+# e = edit settings of procedure
+
+def register_device(name=None, description=''):
+    db = current.db
+    devices = db.devices
+    permission = db.user_permission
+    device_id = devices.insert(name=name, user_id=auth.user_id, description=description)
+    permission.insert(perm_user_email=auth.user.email, device_id=device_id, procedure_id=None, perm_type='a')
+
+
+# Share one device to multiple users with one permission type
+# return emails which are not successfully shared.
+def share_device(device_id, user_email=[], type='v', procedure_id=None):
+    db = current.db
+    permission = db.user_permission
+    fail_email = []
+    for email in user_email:
+        # Does the email exist in registered user table?
+        # if it does not exist this permission adding operation is denied
+        if db(db.auth_user.email == email) is None:
+            fail_email.append(email)
+        else:
+            # Does permission table contain this record?
+            p = db((db.user_permission.perm_user_email == email) &
+                   (db.user_permission.device_id == device_id) &
+                   (db.user_permission.procedure_id == procedure_id)).select().first()
+
+            if p is None:
+                # if no record stored in permission table we insert new permission record
+                permission.insert(perm_user_email=email, device_id=device_id, procedure_id=procedure_id, perm_type=type)
+            else:
+                # update the selected row
+                p.update(perm_user_email=email, device_id=device_id, procedure_id=procedure_id, perm_type=type)
+    return fail_email
+
+
+def can_view_procedure(user_email, device_id, procedure_id):
+    db = current.db
+
+    # Does the user have generic permission to the whole device?
+    p = db((db.user_permission.perm_user_email == user_email) &
+           (db.user_permission.device_id == device_id) &
+           (db.user_permission.procedure_id == None)).select().first()
+    if p.perm_type in {'v', 'e', 'a'}:
```

Comment:
I think rather than sprinkling your code with these 'v', 'e', 'a', etc, you should create a module that treats permissions as elements of a partial order, and implement: 

permissions_max(x, y)
permissions_entails(x, y) (if you have x, does it mean you have y also?)
etc. 

So you could ask: 
if permission_entails(p.perm_type, 'v'):

This way, if we add more permissions than 'v', 'e', 'a', you don't have to redo all your code.